src/algebra/Module/basic.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2019 Robert A. Spencer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert A. Spencer
+-/
+import algebra.module
+import algebra.punit_instances
+import category_theory.types
+import linear_algebra.basic
+open category_theory
+
+universe u
+
+variables (R : Type u) [ring R]
+
+/-- An example of a module for any ring is the zero module -/
+instance module.zero_module : module R punit := module.of_core $
+  by refine {
+    smul := λ _ _, punit.star,
+    .. punit.comm_ring, ..};
+    intros; exact subsingleton.elim _ _
+
+/-- The category of R-modules and their morphisms. -/
+structure Module :=
+  (carrier : Type)
+  (prop_add_comm_group : add_comm_group carrier)
+  (prop_module : module R carrier)
+
+namespace Module
+  instance : has_coe_to_sort (Module R) :=
+    { S := Type, coe := Module.carrier}
+
+  instance add_comm_group (M : Module R) : add_comm_group M := M.prop_add_comm_group
+  instance R_module (M : Module R) : module R M := M.prop_module
+
+  def of (X : Type) [h₁ : add_comm_group X] [h₂ : module R X] : Module R := ⟨ X , h₁ , h₂⟩
```

Comment:
If you define the property fields of `Module` with square brackets, you can just write `:= ⟨X⟩` here, and not even have to give names to h1 and h2.

Reply:
This change makes `of` pretty much redundant with the constructor though. `bundled` is set up with the structure as a `()` argument to make it more convenient to provide the structure explicitly when that is what you want to do. Probably best is to match `bundled`, with `(foo : bar . tactic.apply_instance)` for each of the two "structure" fields.

Reply:
I agree that `of` becomes redundant with `⟨ R, X⟩`.  I'm leaning towards removing it. I'm not quite sure what you are referring to with `()` arguments etc?

Reply:
Well, it's not exactly redundant with `⟨R, X⟩`, because Lean won't generally know the expected type, so won't be able to typecheck this. But I agree that writing `Module.mk R M` is no worse (perhaps better!) than `Module.of R M`. It would be nice to eventually aim for uniformity across all our bundled categories.